### ISSUE TITLE
Add RTL support to ImageMagick

### ISSUE THREAD

**Issue by Fahad-Alsaidi** [ImageMagick/ImageMagick#74]
I hope you can add RTL language support to ImageMagick. This pull request 
https://github.com/ImageMagick/ImageMagick/pull/69
does the needful to enable RTL support in ImaeMagick but unfortunately it seems doesn't fully accepted.

**Comment by urban-warrior** [comment]
The framework for support is there.  ImageMagick is designed to validate delegate libraries @ configure time. That is best practices and why we did not include the libraqm source in ImageMagick.   If you build the libraqm library, the configure script should detect it and call the libraqm API.  Feel free to test and ensure this works.  If any additional patches are required, create another pull request with your mods.

**Comment by khaledhosny** [comment]
The idea is that [libraqm](https://github.com/HOST-Oman/libraqm) is not really a shared library, just a few convenience functions that we use to avoid rewriting the same code repeatedly. From ImageMagic point of view it should be seen as just a source code file that encapsulates the CTL logic, the actual libraries doing the heavy lifting are FriBiDi and HarfBuzz and they are checked at configure time in #69 (but even the configure checks were not committed in master AFAICT).

**Comment by urban-warrior** [comment]
> ImageMagic point of view it should be seen as just a source code file that encapsulates the CTL logic

We require a separate library.  That way libraqm is always maintained separately from ImageMagick, we don't need to check for source code updates, nor concern ourselves with bug fixes.  We are willing to move forward if libraqm is maintained as a static / shared library, otherwise you are welcome to maintain a separate patch that users can apply against the ImageMagick source distribution.

**Comment by khaledhosny** [comment]
Would it make a difference if we moved the code in `raqm.c` inside `annotate.c` (under the `#ifdef`s), because I don’t think a ~700 lines library is worth the trouble, since all it does is gluing FriBiDi and HarfBuzz together and is unlikely to need much updates apart from regular code maintenance.

**Comment by urban-warrior** [comment]
ImageMagick is modular as a best practice to keep MagickCore less bloated.  We are not keen on placing raqm.c inside annotate.c for that reason.  If there were no alternatives we would give it stronger consideration.   If you don't feel like supporting a library, we might create the library ourselves, but given the demands on our schedule, it will likely be sometime next year.

**Comment by khaledhosny** [comment]
To give some context, we were adding CTL support to [SDL_ttf](https://bugzilla.libsdl.org/show_bug.cgi?id=3211) and decided to use HarfBuzz + FriBiDi as the other alternatives are too high level and need much bigger dependencies, but HarfBuzz and FriBiDi are low level libraries and need a bit of plumbing to use them so we wrote a few simple helper functions. When we worked on ImageMagic we realized that we need those same helper functions as well, so we spun them off to a separate file and repository to easily reuse the code. I don’t feel adding an extra third library as a dependency is justified as it adds extra complexity and we will need to convince Linux packagers to package yet another library etc. As an alternative, would using a git submodule be good enough for you?
